### PR TITLE
Backoff retry 

### LINKED ISSUE #562: Backoff Retry During Network Errors
Would it be possible to have Go-Redis retry connections when maxretries is set with a backoff sleep interval? (https://github.com/jpillora/backoff) 

During intervals when my network is overloaded, it is not using backoff and making the overloaded network connection even worst by retrying immediately and producing other errors.

I'd be happy to submit a PR if this is a feature the community/maintainer wants to see in this library.

### PR DESCRIPTION
**Pull request by jonmchan**
Fixes #562 .

Notes:

* Made ```max number of clients reached``` a retryable error
* Backoff Max Time is configurable via config options

Please let me know if I need to adjust anything to make this PR mergable... Thanks!

### CODE REVIEW

**Comment by vmihailenco** [comment]
There is no point in retrying `ErrPoolTimeout`. If you are willing to wait more for a connection then you can just increase `PoolTimeout` * number of retries.

File: internal/errors.go [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+
 const Nil = RedisError("redis: nil")
 
 type RedisError string
 
 func (e RedisError) Error() string { return string(e) }
 
 func IsRetryableError(err error) bool {
-	return IsNetworkError(err)
+	return IsNetworkError(err) || err == ErrPoolTimeout || err == RedisError("ERR max number of clients reached")
```
Comment (vmihailenco) [inline internal/errors.go]: `err.Error() == "ERR max number of clients reached"` reads better and does not allocate.

File: options.go [hunk 2]
```diff
@@ -34,6 +34,10 @@ type Options struct {
 	// Default is to not retry failed commands.
 	MaxRetries int
 
+	// Retry using exponential backoff wait algorithm between each retry
+	// Default is 512 milliseconds; set to -1 to disable
+	MaxBackoffRetryTimeout time.Duration
```
Comment (vmihailenco) [inline options.go]: I think this should be called just `MaxRetryBackoff`.

File: options.go [hunk 3]
```diff
@@ -108,6 +112,11 @@ func (opt *Options) init() {
 	if opt.IdleCheckFrequency == 0 {
 		opt.IdleCheckFrequency = time.Minute
 	}
+	if opt.MaxBackoffRetryTimeout == 0 {
```
Comment (vmihailenco) [inline options.go]:
```
switch opt.MaxRetryBackoff:
case -1:
case 0:
```

File: redis.go [hunk 4]
```diff
@@ -96,9 +96,17 @@ func (c *baseClient) WrapProcess(fn func(oldProcess func(cmd Cmder) error) func(
...
 	for i := 0; i <= c.opt.MaxRetries; i++ {
+		if i > 0 {
+			time.Sleep(internal.RetryBackoff(i, c.opt.MaxBackoffRetryTimeout))
+		}
+
 		cn, _, err := c.conn()
 		if err != nil {
 			cmd.setErr(err)
+
+			if err != nil && internal.IsRetryableError(err) {
```
Comment (vmihailenco) [inline redis.go]: Remove `err != nil && `, because we check for nil 2 lines above.

File: redis_test.go [hunk 5]
```diff
@@ -156,6 +156,51 @@ var _ = Describe("Client", func() {
...
 
+
+	It("should retry with backoff", func() {
+		Expect(client.Close()).NotTo(HaveOccurred())
+
+		// use up all the available connections to force a fail
+		connectionHogClient := redis.NewClient(&redis.Options{
+			Addr:       redisAddr,
+			MaxRetries: 1,
+		})
```
Comment (vmihailenco) [inline redis_test.go]: defer connectionHogClient.Close()

File: redis_test.go [hunk 6]
```diff
@@ -156,6 +156,51 @@ var _ = Describe("Client", func() {
...
+		startRetry := time.Now()
+		err = clientRetry.Ping().Err()
+		Expect(err).To(HaveOccurred())
+		elapseRetry := time.Since(startRetry)
+
+		Expect(elapseRetry > elapseNoRetry).To(BeTrue())
+
+		clientRetry.Close()
+		clientNoRetry.Close()
+		connectionHogClient.Close()
```
Comment (vmihailenco) [inline redis_test.go]: this should be replaced with `defer clientRetry.Close()` etc

**Comment by vmihailenco** [comment]
Overall looks fine, but please fix the build

File: options.go (hunk 3 above)
Comment (jonmchan) [inline options.go]: Do you want me to change https://github.com/Malwarebytes/redis/blob/a3e42eccb641f8f4bcee2cc1a71ab23593de9564/options.go#L101 and https://github.com/Malwarebytes/redis/blob/a3e42eccb641f8f4bcee2cc1a71ab23593de9564/options.go#L96 too? I was just following the current style of the file. This syntax is used for the other timeouts as well.

File: redis.go (hunk 4 above)
Comment (jonmchan) [inline redis.go]: Should the ```err != nil &&``` check be removed from https://github.com/go-redis/redis/pull/570/files#diff-bbc64daf788a83d9a6daf7056d49d2a8R117 as well? Also, I noticed that ```err = readReply(cn)``` (https://github.com/go-redis/redis/pull/570/files#diff-bbc64daf788a83d9a6daf7056d49d2a8R124) is missing a corresponding ```cmd.setErr(err)```. Is that necessary to run that function when readReply returns error?

File: internal/errors.go [hunk 7]
```diff
@@ -1,19 +1,24 @@
...
 import (
+	"errors"
 	"io"
 	"net"
 	"strings"
 )
 
+
+var ErrClosed = errors.New("redis: client is closed")
+var ErrPoolTimeout = errors.New("redis: connection pool timeout")
```
Comment (vmihailenco) [inline internal/errors.go]: remove these 2 vars?

**Comment by vmihailenco** [comment]
I will finish this after merging. Thanks for contributing this!

**Comment by vmihailenco** [comment]
>Also, I noticed that err = readReply(cn) (https://github.com/go-redis/redis/pull/570/files#diff-bbc64daf788a83d9a6daf7056d49d2a8R124) is missing a corresponding cmd.setErr(err). Is that necessary to run that function when readReply returns error?

It is confusing and requires cleanup, but readReply is reponsible to set `cmd.value` and `cmd.err`.

**Pull request merged**